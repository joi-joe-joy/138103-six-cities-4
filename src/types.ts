export enum HouseType {
  APARTMENT = `apartment`,
  ROOM = `room`,
  HOUSE = `house`,
  HOTEL = `hotel`
};

export enum PlaceCardType {
  CITIES = `cities`,
  NEAR = `near`,
  FAVORITES = `favorites`,
  PROPERTY = `property`
};

export enum PageType {
  PROPERTY = `property`,
  MAIN = `main`,
  LOGIN = `login`,
  FAVORITES = `favorites`,
  FAVORITES_EMPTY = `favorites_empty`
};

export enum SortType {
  POPULAR = `popular`,
  TO_HIGH = `to-high`,
  TO_LOW = `to-low`,
  TOP_RATED = `top-rated`
};

export enum AuthStatus {
  NO_AUTH = `NO_AUTH`,
  AUTH = `AUTH`,
};

export interface Location {
  latitude: number,
  longitude: number,
  zoom: number
}

export interface City {
  location: Location,
  name: string
}

export interface User {
  id: number,
  avatarUrl: string,
  name: string,
  isPro: boolean
}

export type AuthInfo = User & {
  email: string
}

export interface Offer {
  id: number,
  title: string,
  maxAdults: number,
  description: string,
  isPremium: boolean,
  isFavorite: boolean,
  images: Array<string>,
  previewImage: string,
  goods: Array<string>,
  price: number,
  rating: number,
  bedrooms: number,
  type: HouseType.APARTMENT | HouseType.ROOM | HouseType.HOUSE | HouseType.HOTEL,
  host: User,
  location: Location,
  city: City,
};

export interface Comment {
  comment: string,
  date: string,
  id: number,
  rating: number,
  user: User
}
