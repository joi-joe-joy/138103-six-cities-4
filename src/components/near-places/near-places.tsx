import * as React from "react";
import {connect} from 'react-redux';
import {Offer, PlaceCardType} from "../../types";
import {getNearbyOffers} from "../../reducer/data/selectors";
import {Operation as DataOperation} from "../../reducer/data/data";
import PlacesList from "../places-list/places-list";

interface Props {
  nearbyOffers: Offer[];
  onLoadNearOffers: (id: number) => void;
  offerId: number;
}

const NearPlaces: React.FC<Props> = (props: Props) => {
  const {nearbyOffers, offerId, onLoadNearOffers} = props;

  React.useEffect(() => {
    if (offerId) {
      onLoadNearOffers(offerId);
    }
  }, [offerId, onLoadNearOffers]);

  if (!nearbyOffers) {
    return null;
  } else {
    return (
      <div className="container">
        <section className="near-places places">
          <h2 className="near-places__title">Other places in the neighbourhood</h2>
          <PlacesList
            type={PlaceCardType.NEAR}
            offers={nearbyOffers}
          />
        </section>
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
  nearbyOffers: getNearbyOffers(state),
});

const mapDispatchToProps = (dispatch) => ({
  onLoadNearOffers(offerId) {
    dispatch(DataOperation.loadNearbyOffers(offerId));
  }
});

export {NearPlaces};
export default connect(mapStateToProps, mapDispatchToProps)(NearPlaces);
