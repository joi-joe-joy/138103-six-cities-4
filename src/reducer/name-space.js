export const NameSpace = {
  DATA: `DATA`,
  USER: `USER`,
  PLACE: `PLACE`
};
