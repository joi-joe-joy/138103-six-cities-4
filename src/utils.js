
export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const noop = () => {
  // do nothing for test
};

// Checking nested objects
const processValue = (value) => {
  if (typeof value !== `object` || value === null) {
    return value;
  } else {
    return Array.isArray(value)
      ? value.map(processValue)
      : renameKeys(value);
  }
};

// Rename Object Keys from Snake Case to Camel Case
export const renameKeys = (object) => {
  return Object.fromEntries(
      Object.entries(object)
        .map(([key, value]) => [
          key.replace(/_(.)/g, (word) => word[1].toUpperCase()),
          processValue(value)
        ])
  );
};
