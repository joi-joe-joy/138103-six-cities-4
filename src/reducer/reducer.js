import {combineReducers} from "redux";
import {reducer as data} from "./data/data";
import {reducer as place} from "./place/place";
import {reducer as user} from "./user/user";
import {reducer as comments} from "./comments/comments";
import {reducer as favorite} from "./favorite/favorite";
import {NameSpace} from "./name-space";

export default combineReducers({
  [NameSpace.DATA]: data,
  [NameSpace.PLACE]: place,
  [NameSpace.USER]: user,
  [NameSpace.COMMENTS]: comments,
  [NameSpace.FAVORITE]: favorite,
});
