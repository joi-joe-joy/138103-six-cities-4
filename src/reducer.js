import {extend} from "./utils.js";
import offers from "./mocks/offers.js";

let initCity = offers[0].city;
let initOffers = offers.filter((offer) => offer.city === initCity);

// Получаем все значения городов
let cities = offers.map((offer)=>offer.city);
// Получаем уникальные значения городов
let citiesList = Array.from(new Set(cities));

const initialState = {
  city: initCity,
  offersCityList: initOffers,
  offers,
  citiesList
};

const ActionType = {
  CHANGE_CITY: `CHANGE_CITY`,
  GET_OFFERS_CITY_LIST: `GET_OFFERS_CITY_LIST`,
};

const ActionCreator = {
  changeCity: (city) => {
    let changedCity = city ? city : offers[0].city;

    return {
      type: ActionType.CHANGE_CITY,
      payload: changedCity
    };
  },
  getOffersCityList: (city) => {
    let offersCityList = offers.filter((offer) => offer.city === city);

    return {
      type: ActionType.GET_OFFERS_CITY_LIST,
      payload: offersCityList
    };
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_CITY:
      return extend(state, {
        city: action.payload
      });
    case ActionType.GET_OFFERS_CITY_LIST:
      return extend(state, {
        offersCityList: action.payload
      });
  }

  return state;
};

export {reducer, ActionType, ActionCreator};
