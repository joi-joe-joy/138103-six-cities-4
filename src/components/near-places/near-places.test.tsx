import * as React from "react";
import * as renderer from "react-test-renderer";
import {BrowserRouter} from "react-router-dom";
import {NearPlaces} from "./near-places";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import {NameSpace} from "../../reducer/name-space";
import {Offer, HouseType} from "../../types";
import {noop} from '../../utils';

const mockStore = configureStore([]);

const offers: Offer[] = [{
  bedrooms: 2,
  city: {
    name: `Paris`,
    location: {
      latitude: 48.85661,
      longitude: 2.351499,
      zoom: 13
    }
  },
  description: `Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.`,
  goods: [`Air conditioning`, `Laptop friendly workspace`, `Baby seat`, `Fridge`, `Breakfast`, `Washer`, `Washing machine`, `Dishwasher`, `Coffee machine`, `Towels`],
  host: {id: 25, name: `Angelina`, isPro: true, avatarUrl: `img/avatar-angelina.jpg`},
  id: 1,
  images: [`https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg`],
  isFavorite: false,
  isPremium: true,
  location: {
    latitude: 48.865610000000004,
    longitude: 2.350499,
    zoom: 16
  },
  maxAdults: 8,
  previewImage: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/5.jpg`,
  price: 397,
  rating: 3.6,
  title: `Penthouse, 4-5 rooms + 5 balconies`,
  type: HouseType.HOTEL
}];

it(`Render ReviewBox correctly with nearby Offers`, () => {
  const store = mockStore({
    [NameSpace.DATA]: {
      nearbyOffers: offers
    },
    [NameSpace.USER]: {
      authorizationStatus: `AUTH`
    },
  });

  const tree = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <NearPlaces
            offerId={1}
            nearbyOffers={offers}
            onLoadNearOffers={noop}
          />
        </BrowserRouter>
      </Provider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it(`Render ReviewBox correctly without nearby Offers`, () => {
  const store = mockStore({
    [NameSpace.DATA]: {
      nearbyOffers: []
    },
    [NameSpace.USER]: {
      authorizationStatus: `AUTH`
    },
  });

  const tree = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <NearPlaces
            offerId={1}
            nearbyOffers={[]}
            onLoadNearOffers={noop}
          />
        </BrowserRouter>
      </Provider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
