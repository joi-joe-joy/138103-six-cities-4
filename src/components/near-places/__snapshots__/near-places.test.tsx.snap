// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ReviewBox correctly with nearby Offers 1`] = `
<div
  className="container"
>
  <section
    className="near-places places"
  >
    <h2
      className="near-places__title"
    >
      Other places in the neighbourhood
    </h2>
    <div
      className="places__list near-places__list"
    >
      <article
        className="place-card near-places__card"
        onMouseOver={[Function]}
      >
        <div
          className="place-card__mark"
        >
          <span>
            Premium
          </span>
        </div>
        <div
          className="place-card__image-wrapper near-places__image-wrapper"
        >
          <a
            href="#"
          >
            <img
              alt="Penthouse, 4-5 rooms + 5 balconies"
              className="place-card__image"
              height="200"
              src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/5.jpg"
              width="260"
            />
          </a>
        </div>
        <div
          className="place-card__info"
        >
          <div
            className="place-card__price-wrapper"
          >
            <div
              className="place-card__price"
            >
              <b
                className="place-card__price-value"
              >
                €
                397
              </b>
              <span
                className="place-card__price-text"
              >
                / night
              </span>
            </div>
            <button
              className="button place-card__bookmark-button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="18"
                style={
                  Object {
                    "fill": "#000000",
                  }
                }
                width="17"
              />
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="place-card__rating rating"
          >
            <div
              className="place-card__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "80%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
          </div>
          <h2
            className="place-card__name"
          >
            <a
              href="/offer/1"
              onClick={[Function]}
            >
              Penthouse, 4-5 rooms + 5 balconies
            </a>
          </h2>
          <p
            className="place-card__type"
          >
            Hotel
          </p>
        </div>
      </article>
    </div>
  </section>
</div>
`;

exports[`Render ReviewBox correctly without nearby Offers 1`] = `
<div
  className="container"
>
  <section
    className="near-places places"
  >
    <h2
      className="near-places__title"
    >
      Other places in the neighbourhood
    </h2>
    <div
      className="places__list near-places__list"
    />
  </section>
</div>
`;
